File: algorithms/c++/572-subtree-of-another-tree.cpp
// https://leetcode.com/problems/subtree-of-another-tree/
#include <bits/stdc++.h>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
}

class Solution {
public:
    bool checkSubtree(TreeNode* root, TreeNode* subRoot) {
        if (!root || !subRoot) {
            if (root == nullptr && subRoot == nullptr)
                return true;
            return false;
        }
        if (root->val != subRoot->val)
            return false;
        return checkSubtree(root->left, subRoot->left) && checkSubtree(root->right, subRoot->right);
    }
    
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if (root->val == subRoot->val) {
            if (checkSubtree(root, subRoot))
                return true;
        }
        return (root->left && isSubtree(root->left, subRoot)) || (root->right && isSubtree(root->right, subRoot));
    }
};

int main() {

    return 0;
}

File: algorithms/python/572-subtree-of-another-tree.py
# https://leetcode.com/problems/subtree-of-another-tree/

from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def checkSubtree(self, node: Optional[TreeNode], subNode: Optional[TreeNode]) -> bool:
        if node is None and subNode is None:
            return True
        if node is None or subNode is None:
            return False
        if node.val != subNode.val:
            return False
        return self.checkSubtree(node.left, subNode.left) and self.checkSubtree(node.right, subNode.right)
    
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if root is None:
            return False
        if root.val == subRoot.val and self.checkSubtree(root, subRoot):
            return True
        return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)
